0.1 dev

Ikstar opened the pull request:
Address #4 

#### Added
- Updated ReadMe with Usage details
- Add Changelog
- Populate YourMembershipServiceProvider for Laravel
- Add minimum-stability for composer.json
- Added Generic YourMembershipException
#### Fixed
- Recursive XML Packaging of API Call Arguments
#### Changed
- Rename YourMembershipException to YourMembershipApiException
#### Removed
- Carbon\Carbon Dependency (unused)

Commit:
Add Service Provider

Commit:
Add support for recursive argument

Add Multiple Exception Types

Commit:
Update Required Packages and Stability

Commit:
Update Documentation

Commit:
clean up

Commit:
fix typos

Commit:
Fix Provider typos

chefboyarsky commented on src/Core/Request.php:
```diff
@@ -85,13 +85,39 @@ public function createCallPayload(string $method, array $arguments) : \SimpleXML
-            $call->addChild($key, $value);
-        }
+
+        $call = $this->sxmlAddChildrenRecursive($call,$arguments);
```
You may want to wrap this with a try/catch since it can throw a YourMembershipException

chefboyarsky commented on src/Core/Request.php:
```diff
@@ -85,13 +85,39 @@ public function createCallPayload(string $method, array $arguments) : \SimpleXML
+     * @param  SimpleXMLElement         $root      Root XML Node
+     * @param  array                    $arguments Array of Arguments to be added to XML Node
+     * @return SimpleXMLElement                    Resulting XML Tree
+     */
```
@throws

chefboyarsky commented on src/Exceptions/YourMembershipException.php:
```diff
@@ -0,0 +1,13 @@
+ * Used typically for errors dealing with XML Parsing Issues
+ *
+ */
+class YourMembershipException extends \Exception
```
Why the superclass? The difference between YourMembershipException and YourMembershipApiException isn't clear to me. From the perspective of a client of your package, they are all sort of the same type of thing, right? Maybe there is some purpose internal to this package.

Ikstar replied:
As we discussed it's a generic Parent class so external libraries can be aware of YourMembership library exceptions

chefboyarsky commented on src/YourMembershipServiceProvider.php:
```diff
@@ -20,12 +23,21 @@ class YourMembershipServiceProvider extends ServiceProvider
 	{
-		$this->package('phone2action/ym-api');
+		$this->publishes([
+		$this->packageConfigPath => config_path('yourmembership.php'),
```
Indentation is weird here

chefboyarsky commented on src/YourMembershipServiceProvider.php:
```diff
@@ -35,7 +47,7 @@ public function boot()
 	public function provides()
 	{
-		return array();
+		return array(YourMembershipClient::class);
```
[] syntax

chefboyarsky commented on tests/unit/Core/ResponseTest.php:
```diff
@@ -110,7 +110,7 @@ public function testToArrayWithError() {
 
         //Execute and Verify
-        $this->tester->expectException(YourMembershipException::class, function() use ($response)
+        $this->tester->expectException(YourMembershipApiException::class, function() use ($response)
```
I haven't used this syntax for exceptions before. I usually use the @expectException PHPUnit annotation on the test method. It may be less code. Syntax is like

@expectException YourMembershipApiException

and then in the test, you'd just call:

$a = $response->toArray()



I'd lean towards that, but it's a matter of preference if it's working.

Ikstar replied:
This is the way Codeception has the assertions setup.

chefboyarsky commented on tests/unit/Core/YourMembershipExceptionTest.php:
```diff
@@ -22,8 +22,10 @@ protected function _after()
+        $exception = new YourMembershipApiException('Message', 999, 'TestMethod');
 
+        //verify
         echo $exception;
```
It would be good to see an assertion in this test. Can you call toString() on it explicitly and assert the value is what you'd expect?

chefboyarsky commented on tests/unit/Core/RequestTest.php:
```diff
@@ -57,14 +77,31 @@ public function testCreateCallPayLoad()
 
+    public function testCreateCallPayLoadWithException()
+    {
+        $this->tester->expectException(YourMembershipException::class, function()
```
See other comment about @expectException

chefboyarsky reviewed (changes requested):
Mostly looking good. A little cleanup to do

jonathanwkelly commented on composer.json:
```diff
@@ -19,9 +19,10 @@
+    "minimum-stability": "dev",
+    "prefer-stable": true,
     "require": {
         "php": ">=7.0.0",
```
Could we get away with 5.6.4 ?

Ikstar replied:
No 7.0 is required for type hinting

Commit:
Move the exception up

Add YourMembershipRequestException
Rename YourMembershipApiException to YourMembershipResponseException

Commit:
Update tests for Request and Response for new Exceptions

Commit:
Fix ServiceProvider

Commit:
Update changelog

Commit:
Bump Version

scrutinizer-auto-fixer committed:
Scrutinizer Auto-Fixes

This commit consists of patches automatically generated for this project on https://scrutinizer-ci.com

Commit:
remove config from code coverage

Commit:
fix exclude path

chefboyarsky commented on src/Core/Request.php:
```diff
@@ -85,13 +86,37 @@ public function createCallPayload(string $method, array $arguments) : \SimpleXML
+        try {
+            $call = $this->sxmlAddChildrenRecursive($call, $arguments);
+        } catch (\Exception $e) {
+            throw new YourMembershipRequestException($e->getMessage(), $e->getCode(), $method, $arguments, $e);
```
Something smells about passing in the exception and its properties. I could see it being useful for a generic exception that people want to give messages to as params, but for a YourMembershipRequestException, it's likely that just passing in $e and pulling the message/code off may be enough

chefboyarsky commented on src/Core/Request.php:
```diff
@@ -85,13 +86,37 @@ public function createCallPayload(string $method, array $arguments) : \SimpleXML
+     * @date   2017-01-12
+     * @param  \SimpleXMLElement         $root      Root XML Node
+     * @param  array                    $arguments Array of Arguments to be added to XML Node
+     * @return \SimpleXMLElement                    Resulting XML Tree
```
The try/catch around the call to this method in createCallPayload suggests that this method can throw exceptions. If so, it needs an annotation. I don't see it throwing any exceptions, though -- do the children throw exceptions? If not, we wouldn't need a try/catch in createCallPayload. If so (more likely), do we want to be catching some of them in this method, and do those children (e.g. sxmlAppend) need @throws docblocks?

Ikstar replied:
```$child = new \SimpleXMLElement(sprintf('<%s></%s>', $key, $key)); ```
This line throws the exception if the $key argument is not valid for XML generation. Perhaps checking if it is a valid $key and throw the exception there, in this case we don't catch Previous exception.

However validating whether a key can make a valid XML node sounds painful and quite a bit more informative when we fail on attempting to with extra detail

Ikstar replied:
Added @throws annotation

Commit:
Address Review Items
